src/computedevs.jl

```diff
@@ -0,0 +1,80 @@
+# adaptors for converting abstract arrays to Base.Array
+
+
+"""
+    abstract type AbstractComputingDevice
+
+Supertype for CPU and GPU computing devices.
+
+`adapt(dev::AbstractComputingDevice, x)` adapts `x` for `dev`.
+"""
+abstract type AbstractComputingDevice end
+export AbstractComputingDevice
+
+
+"""
+    struct ComputingDeviceIndependent <: AbstractComputingDevice
+
+`get_computing_device(x) === ComputingDeviceIndependent()` indicates
+that `x` is not tied to a specific computing device. This typically
+means that x is a statically allocated object.
+"""
+struct ComputingDeviceIndependent <: AbstractComputingDevice end
+export ComputingDeviceIndependent
+
+
+"""
+    struct CPUDevice <: AbstractComputingDevice
+
+`CPUDevice()` is the default CPU device.
+"""
+struct CPUDevice <: AbstractComputingDevice end
+export CPUDevice
+
+
+
+"""
+    abstract type AbstractGPUDevice <: AbstractComputingDevice
+
+Supertype for GPU computing devices.
+"""
+abstract type AbstractGPUDevice <: AbstractComputingDevice end
+export AbstractGPUDevice
+
+
+
+const _incompatible_devs = ArgumentError("Incompatible computing devices")
+
+select_computing_device(a::ComputingDeviceIndependent, ::ComputingDeviceIndependent) = a
+select_computing_device(a::ComputingDeviceIndependent, b::AbstractComputingDevice) = b
+select_computing_device(a::AbstractComputingDevice, b::ComputingDeviceIndependent) = a
+
+select_computing_device(a::CPUDevice, ::CPUDevice) = a
+select_computing_device(a::CPUDevice, b::AbstractGPUDevice) = a
+select_computing_device(a::AbstractGPUDevice, b::CPUDevice) = b
+select_computing_device(a::AbstractGPUDevice, b::AbstractGPUDevice) = (a === b) ? a : throw(_incompatible_devs)
```

Comment:
Note that throwing a nonspecific exception on error makes it impossible to implement an efficient fallback safely. It would be more composable if it were returning some kind of value ("bottom") representing that an appropriate device is undecidable.

Reply:
Oh, yes, I like it - we recently took that approach in InverseFunctions and ChangesOfVariables. If the generated function (see above) goes, we may not need `select_computing_device` at all anymore, at least for now, but we could use a bottom value in the default implementation of `get_computing_device` (see comment by @tkf  above).

Reply:
I'm adding a bottom value